File: golang/abi/typeFormula.go
package abi

import (
	"fmt"
	"strings"
)

type TypeFormula struct {
	Name           string
	TypeParameters []*TypeFormula
}

func NewTypeFormula(name string, typeParameters []*TypeFormula) *TypeFormula {
	return &TypeFormula{
		Name:           name,
		TypeParameters: typeParameters,
	}
}

func (t *TypeFormula) String() string {
	if len(t.TypeParameters) > 0 {
		var typeParameters []string
		for _, typeParameter := range t.TypeParameters {
			typeParameters = append(typeParameters, typeParameter.String())
		}
		return fmt.Sprintf("%s<%s>", t.Name, strings.Join(typeParameters, ", "))
	} else {
		return t.Name
	}
}

File: typescript/src/typeFormula.ts
export class TypeFormula {
    name: string;
    type_parameters: TypeFormula[];

    constructor(name: string, type_parameters: TypeFormula[]) {
        this.name = name;
        this.type_parameters = type_parameters;
    }

    toString(): string {
        if (this.type_parameters.length > 0) {
            const type_parameters = this.type_parameters.map((type_parameter) => type_parameter.toString()).join(", ");
            return `${this.name}<${type_parameters}>`;
        } else {
            return this.name;
        }
    }
}
